src/cargo/sources/path.rs

```diff
@@ -411,6 +464,151 @@ impl<'gctx> PathSource<'gctx> {
         }
     }
 
+    /// Lists files relevant to building this package inside this source by
+    /// consulting traversing the git working tree, while avoiding ignored files.
+    ///
+    /// This looks into Git submodules as well, resolving them to individual files.
+    /// Symlinks to directories will also be resolved, which may include `.git` folders
+    /// in entirety if the link points to a submodule.
+    fn list_files_gix(
+        &self,
+        pkg: &Package,
+        repo: &gix::Repository,
+        filter: &dyn Fn(&Path, bool) -> bool,
+    ) -> CargoResult<Vec<PathBuf>> {
+        warn!("list_files_gix {}", pkg.package_id());
+        let options = repo
+            .dirwalk_options()?
+            .emit_untracked(gix::dir::walk::EmissionMode::Matching)
+            .emit_ignored(None)
+            .emit_tracked(true)
+            .recurse_repositories(false)
+            .symlinks_to_directories_are_ignored_like_directories(true)
+            .emit_empty_directories(false);
+        let index = repo.index_or_empty()?;
+        let root = repo
+            .work_dir()
+            .ok_or_else(|| anyhow::format_err!("can't list files on a bare repository"))?;
+        assert!(
+            root.is_absolute(),
+            "BUG: paths used internally are absolute, and the repo inherits that"
+        );
+
+        let pkg_path = pkg.root();
+        let mut target_prefix;
+        let repo_relative_pkg_path = pkg_path.strip_prefix(root).ok().unwrap_or(Path::new(""));
+
+        let pathspec = {
+            let mut include = gix::path::to_unix_separators_on_windows(gix::path::into_bstr(
+                repo_relative_pkg_path,
+            ));
+
+            target_prefix = include.clone();
+            target_prefix.to_mut().push_str(if include.is_empty() {
+                "target/"
+            } else {
+                "/target/"
+            });
+
+            // The `target` directory is never included if it is in the package root.
+            let exclude = {
+                let mut pattern = include.clone();
+                pattern
+                    .to_mut()
+                    .insert_str(0, if include.is_empty() { ":!" } else { ":!/" });
```

Comment:
This pathspec prefix is short for  'exclude from top', with the 'top' being required to avoid the CWD to affect the pathspec.

Reply:
What is the `:` syntax and where can I read more about it?

Reply:
Oh, is it a path separator?  Why are we using strings with separators rather than `Vec`?

Reply:
The `pathspec` syntax is document in [the Git glossary](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefpathspecapathspec) of all places.

With a top-level `Cargo.toml` file, this code produces two pathspecs: `:/` and `:!/target/`, with a sub-package it would look like `:/sub-package` and `:!/sub-package/target/`.

Reply:
Why is a Rust API using stringly-typed values based on a command-line?

Reply:
Now I see what you are getting at.

I didn't think of innovating here, following the API used by `libgit2` and `git2` respectively. Having strings for pathspecs runs deep, and they are treated more like paths or regex or globs rather than a data structure. [The datastructure itself](https://github.com/Byron/gitoxide/blob/f8ce3d0721b6a53713a9392f2451874f520bc44c/gix-pathspec/src/lib.rs#L84) also isn't particularly well suited to be constructed by hand, as git settings affect it (like case-sensitivity). It's quite complex, but I can imagine to tackle this by adding a trait like `TryIntoPathspec` which then works for strings as well as for hand-created pathspecs.

I see how being able to not have to go through the parser here would be beneficial though, so I have [put this into the roadmap](https://github.com/Byron/gitoxide/commit/3fd668454acece9d218dc7b5200836dce938efbc).

(*Since it's not worse than `git2` which is also stringly typed, I hope this won't have to block this PR.*)

Reply:
No, thank you for the explanation!

For me, it is more of trying to make the intent of the code clearer.  I'll mark this as resolved but have some other thoughts for improving it with what we have now.

Reply:
Hmm, was tempted to encourage `pathspec` being added to the variables but the leading `let pathspec =` helps.  Still easy to miss the structure these variables are intended to uphold.  I feel like I could go either way.

Reply:
It's definitely quite complex code, with a bunch of variables and inter-dependencies to make it all line up. This could already be improved a lit if `format!()` could be used, but since paths are involved, that's not possible :/.
Maybe the pathspec-code can be refactored into something better nonetheless.